File: js/maths/powerof2.js
// problem: Given a positive integer of 'n', determine if the number is a power of 2


//O(logn)
function powerOf2(n) {

    if ( n < 1 ) return false;
    while ( n > 1 ) {
        if ( n % 2 !== 0 ) {
            return false
        }

        n = n / 2;
    }
    return true

}

//console.log(powerOf2(4))

//O(1)
const bitwisePowerOf2 = ( n ) => {
    
    if ( n < 1 ) return false;

    return ( n & ( n - 1 ) ) === 0;
};

console.log( bitwisePowerOf2( 4 ) );
console.log( bitwisePowerOf2( 7 ) );
console.log( bitwisePowerOf2( 8 ) );
console.log(bitwisePowerOf2(16))

File: java/maths/PowerOf2.java
public class PowerOf2 {

    // O(logn)
    static boolean power(int n) {

        if (n < 1)
            return false;

        while (n > 1) {
            if (n % 2 != 0)
                return false;
            n = n / 2;
        }

        return true;
    }
// O(1)
    static boolean bitwisePowerOf2(int n) {
        if (n < 1)
            return false;

        return (n & (n - 1)) == 0;
        
    }
    
    public static void main(String[] args) {
        System.out.println( bitwisePowerOf2( 4 ) );
        System.out.println( bitwisePowerOf2( 7 ) );
        System.out.println( bitwisePowerOf2( 8 ) );
        System.out.println(bitwisePowerOf2(1));
    }
}
